builder/steps/gen-dockerfile/contents/src/detect_setup.ts

```diff
@@ -0,0 +1,167 @@
+ * This function is used to detect information about a Node.js application.
+ * In particular, this function determines whether or not a `package.json`
+ * file is available to allow `npm` or `yarn` to do an install, whether
+ * the package.json file has a start script specified, whether the package.json
+ * has an engines field specified that specifies a Node.js version to use,
+ * and whether `yarn` or `npm` should be used to install dependencies and run
+ * the application.
+ *
+ * @param logger Used to print statements designed for user consumption.
+ * @param fsview Represents a view of the root directory of a Node.js
+ *               application.
+ *
+ * @throws If an `app.yaml` file does not exist in the directory desribed by
+ *         the given {@link Locator}.
+ * @throws If a `package.json` file exists but a problem occurred while
+ *         trying to parse the file.
+ * @throws If the directory described by the {@link Reader} and {@link Locator}
+ *         does not have a start script in the `package.json` file (or the
+ *         `package.json` file doesn't exist) and the directory does not
+ *         contain a `server.js` file.
+ *
+ * @return Information about the Node.js application in question that can
+ *         be used to determine how to install the application's dependencies
+ *         and run the application.
+ */
+export async function detectSetup(logger: Logger, fsview: Reader&Locator): Promise<Setup> {
+  if (!(await fsview.exists(APP_YAML))) {
+    throw new Error(`The file ${APP_YAML} does not exist`);
+  }
+  const config = yaml.safeLoad(await fsview.read(APP_YAML));
+
+  // If nodejs has been explicitly specified then treat warnings as errors.
+  const warn: (m: string) => void =
+      config.runtime ? logger.error.bind(logger) : logger.log.bind(logger);
+
+  logger.log('Checking for Node.js.');
+
+  let canInstallDeps: boolean;
+  let gotScriptsStart: boolean;
+  let nodeVersion: string|undefined;
+  let useYarn: boolean;
+
+  if (!(await fsview.exists(PACKAGE_JSON))) {
+    logger.log('node.js checker: No package.json file.');
+    canInstallDeps = false;
+    gotScriptsStart = false;
+    nodeVersion = undefined;
+    useYarn = false;
+  } else {
+    canInstallDeps = true;
+
+    // Consider the yarn.lock file as present if and only if the yarn.lock
+    // file exists and is not specified as being skipped in app.yaml.
+    let skipFiles = config.skip_files || [];
+    if (!Array.isArray(skipFiles)) {
+      skipFiles = [skipFiles];
+    }
+
+    const yarnLockExists: boolean = await fsview.exists(YARN_LOCK);
+    let yarnLockSkipped = false;
+    skipFiles.forEach((pattern: string) => {
+      yarnLockSkipped = yarnLockSkipped || new RegExp(pattern).test(YARN_LOCK);
+    });
+
+    useYarn = yarnLockExists && !yarnLockSkipped;
+
+    // Try to read the package.json file.
+    let packageJson;
+    const contents = await fsview.read(PACKAGE_JSON);
+
+    try {
+      packageJson = JSON.parse(contents);
```

Comment:
I haven't changed this yet because by not using `require` makes it easier to mock things in the tests.  It makes most sense to address this when I address the use of `exist` to check for file existence.